### PR TITLE
Crd generation

### PR DESCRIPTION
**Pull request by KevinJoiner**
Issue https://github.com/rancher/rancher/issues/39335
These updates are to support the creation of CRDs from API structs instead of our custom CRD defs which we will no longer need to maintain.

Commit 1
The first commit adds the ability to decode a YAML document into an object with JSON decoding. This is needed so we do not have to work with unstructered.Unstuctered Objects and map decoding.

Commit 2
Adds a new BatchCreate that creates CRDs from the apis structs. This is needed to support the new CRD generation in Rancher, which no longer use wrangler crd.CRD

### CODE REVIEW

File: pkg/crd/crd.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create crd '%s': %w", crd.Name, err)
+		}
+		// item was not in initial list but does exist so we will attempt to get the latest resourceVersion
+		existingCRD, err = crdClient.Get(ctx, crd.Name, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get exiting '%s' CRD: %w", crd.Name, err)
+		}
+		// fallthrough and attempt and update
```
Comment (JonCrowther) [inline pkg/crd/crd.go:90]:
```suggestion
		// fallthrough and attempt an update
```

File: pkg/crd/crd.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	// only keep the resource version for the desired object
+	crd.ResourceVersion = existingCRD.ResourceVersion
+	// if the CRD exist attempt to update it
```
Comment (JonCrowther) [inline pkg/crd/crd.go:94]:
```suggestion
	// if the CRD exists attempt to update it
```

File: pkg/crd/crd.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		if err != nil {
+			return fmt.Errorf("failed waiting on CRD '%s': %w", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
```
Comment (JonCrowther) [inline pkg/crd/crd.go:58]: Since this gets all existing CRDs, it should be plural
```suggestion
func getExistingCRDs(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
```

File: pkg/crd/crd.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	if !exist {
+		logrus.Infof("Creating embedded CRD %s", crd.Name)
+		_, err := crdClient.Create(ctx, crd, metav1.CreateOptions{})
+		if err == nil {
+			return nil
+		}
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create crd '%s': %w", crd.Name, err)
+		}
+		// item was not in initial list but does exist so we will attempt to get the latest resourceVersion
```
Comment (JonCrowther) [inline pkg/crd/crd.go:85]: In what situation would this happen? Where `crdClient.List` would not contain something that could be gotten via `crdClient.Get`?

File: pkg/crd/crd.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+	})
+}
+
+func crdIsReady(crd *apiextv1.CustomResourceDefinition) bool {
+	for i := range crd.Status.Conditions {
+		cond := &crd.Status.Conditions[i]
+		switch cond.Type {
+		case apiextv1.Established:
+			if cond.Status == apiextv1.ConditionTrue {
+				return true
```
Comment (JonCrowther) [inline pkg/crd/crd.go:127]: Just want to make sure this is intentional. If there are any `Established` status conditions that succeeded, we want to return true? No matter how many errors there may be?

File: pkg/crd/crd.go (hunk 4 above)
Comment (KevinJoiner) [reply pkg/crd/crd.go:85]: 2 scenarios, both of which don't get hit right now.

1. Allow for a label selector in the List so we do not have to pull every CRD we may not care about.
I will update the BatchCreate call to take one in.

2. If the user does not have list access the logic can still be performed if you have the get verb.
This does not work right now because of your previous comment where I am returning the error.

File: pkg/crd/crd.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+	// ensure each CRD exist
+	for _, crd := range crds {
+		err = ensureCRD(ctx, crd, crdClient, existingCRDs)
+		if err != nil {
+			return err
+		}
+	}
+
+	existingCRDs, err = getExistingCRD(ctx, crdClient)
```
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/crd/crd.go:38]: 1. I forgot to stop returning an error here.
2. The second list is to get the updated versions of the CRDs. After the creation or update of the CRDs K8s will add a status to the resource, letting us know if it is ready or not. We make the second list so we can get that status information and determine if we need to wait for the CRD to be ready with `waitCRD(...)` or not.

File: pkg/crd/crd.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
+// Package crd helps with the installation of CustomResourceDefinitions.
```
Comment (maxsokolovsky) [inline pkg/crd/crd.go:1]: Tiny nit: I'd add a more descriptive or concrete comment, not sure what "helps" really means.

File: pkg/crd/crd.go [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
+	storedCRDs, err := crdClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list crds: %w", err)
+	}
+
+	// convert existingCRDs to a map for faster lookup
+	existingCRDs := map[string]*apiextv1.CustomResourceDefinition{}
```
Comment (maxsokolovsky) [inline pkg/crd/crd.go:65]: You could pre-allocate the capacity of the map.

File: pkg/crd/crd.go (hunk 5 above)
Comment (KevinJoiner) [reply pkg/crd/crd.go:127]: Yes, the `condition.Type` is a unique key for all of the conditions in the list.
You may be curious why we don't return false if the condition is not true. This is because we also want to log if the CRD has a naming conflict for the case below, so we have to loop through the entire list to check for both.

File: pkg/crd/crd.go [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+		if err != nil {
+			return fmt.Errorf("failed to get exiting '%s' CRD: %w", crd.Name, err)
+		}
+		// fallthrough and attempt and update
+	}
+	// only keep the resource version for the desired object
+	crd.ResourceVersion = existingCRD.ResourceVersion
+	// if the CRD exist attempt to update it
+	logrus.Infof("Updating embedded CRD %s", crd.Name)
+	_, err := crdClient.Update(ctx, crd, metav1.UpdateOptions{})
```
Comment (maxsokolovsky) [inline pkg/crd/crd.go:96]: Since you update the CRD in the cluster, would you want to also update it in your in-memory map?

Comment (KevinJoiner) [reply pkg/crd/crd.go:96]: This map is not reused after this, so there is no need to update it.

File: pkg/crd/crd.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+		}
+	}
+
+	existingCRDs, err = getExistingCRD(ctx, crdClient)
+	if err != nil {
+		return fmt.Errorf("unable to check for existing CRDs: %w", err)
+	}
+	// ensure each CRD is ready
+	for _, crd := range crds {
+		if existingCRD, ok := existingCRDs[crd.Name]; ok && crdIsReady(existingCRD) {
```
Comment (maxsokolovsky) [inline pkg/crd/crd.go:44]: I think you already know that the CRD exists in the map, since you'd run `ensure` above?

Comment (KevinJoiner) [reply pkg/crd/crd.go:44]: Correct I added a label selector to the list so there is a chance the CRD does not have that label and is not returned in the list.

File: pkg/yaml/yaml.go [hunk 11]
```diff
@@ -24,15 +26,62 @@ var (
 	cleanContains = []string{
 		"cattle.io/",
 	}
+	buffSize = 4096
```
Comment (maxsokolovsky) [inline pkg/yaml/yaml.go:29]: This may be a constant.

File: pkg/yaml/yaml.go [hunk 12]
```diff
@@ -24,15 +26,62 @@ var (
...
 )
 
+// Unmarshal decodes YAML bytes into document (as defined by
+// the YAML spec) then converting it to JSON via
+// "k8s.io/apimachinery/pkg/util/yaml".YAMLToJSON and then decoding the json in the the v interface{}
 func Unmarshal(data []byte, v interface{}) error {
 	return yamlDecoder.NewYAMLToJSONDecoder(bytes.NewBuffer(data)).Decode(v)
 }
 
+// UnmarshalWithJSONDecoder takes a reader of yaml bytes converts the document or documents to JSON
```
Comment (maxsokolovsky) [inline pkg/yaml/yaml.go:39]: Could you tighten up the comment a bit? I am being so annoying because it's an exported function. 😺

```suggestion
// UnmarshalWithJSONDecoder expects a reader of raw YAML. 
// It converts the document or documents to JSON,
// then decodes the JSON bytes into a slice of values of type T.
// Type T must be a pointer, or the function will panic.
```

File: pkg/yaml/yaml.go [hunk 13]
```diff
@@ -24,15 +26,62 @@ var (
 	cleanContains = []string{
 		"cattle.io/",
 	}
+	buffSize = 4096
 )
 
+// Unmarshal decodes YAML bytes into document (as defined by
+// the YAML spec) then converting it to JSON via
+// "k8s.io/apimachinery/pkg/util/yaml".YAMLToJSON and then decoding the json in the the v interface{}
 func Unmarshal(data []byte, v interface{}) error {
```
Comment (maxsokolovsky) [inline pkg/yaml/yaml.go:35]: Is this function similar to the new one? We just shouldn't use it anymore because it's less safe the new, generic one?

File: pkg/yaml/yaml_test.go [hunk 14]
```diff
@@ -0,0 +1,161 @@
+package yaml
```
Comment (maxsokolovsky) [inline pkg/yaml/yaml_test.go:1]: You could make all tests and subtests run in parallel.

**Review by maxsokolovsky — CHANGES_REQUESTED** [review]

File: pkg/crd/crd_test.go [hunk 15]
```diff
@@ -0,0 +1,308 @@
+// Package crd helps with the installation of CustomResourceDefinitions.
```
Comment (maxsokolovsky) [inline pkg/crd/crd_test.go:1]: Could you please update this package comment as well? This one seems to win when it comes to our godoc - there are two package comments now.

File: pkg/crd/crd.go [hunk 16]
```diff
@@ -0,0 +1,143 @@
...
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clientv1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	waitDuration = 1 * time.Minute
```
Comment (maxsokolovsky) [inline pkg/crd/crd.go:19]: Feel free to ignore this nit - but I would make these constants here and pass them to your waitForCRD function. They seem to be variables just so they can be changed in tests, which is OK, but not ideal.

Comment (KevinJoiner) [reply pkg/crd/crd.go:19]: Since `time.Minute` is technically a struct the compiler won't allow me to make it constant.

File: pkg/crd/crd_test.go (hunk 15 above)
Comment (KevinJoiner) [reply pkg/crd/crd_test.go:1]: missed this one.

File: pkg/crd/crd.go (hunk 16 above)
Comment (maxsokolovsky) [reply pkg/crd/crd.go:19]: It will, since values of type `time.Duration` are really just numbers.

File: pkg/yaml/yaml.go (hunk 12 above)
Comment (KevinJoiner) [reply pkg/yaml/yaml.go:39]: Can do it's a good suggestion 👍🏾

File: pkg/crd/mockCRDClient_test.go [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+
+	gomock "github.com/golang/mock/gomock"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	v10 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	types "k8s.io/apimachinery/pkg/types"
+	watch "k8s.io/apimachinery/pkg/watch"
+)
+
+// MockCustomResourceDefinitionInterface is a mock of CustomResourceDefinitionInterface interface.
+type MockCustomResourceDefinitionInterface struct {
```
Comment (maxsokolovsky) [inline pkg/crd/mockCRDClient_test.go:19]: A general question: why use mocks here? Couldn't you create a fake implementation instead?

File: pkg/crd/crd_test.go [hunk 18]
```diff
@@ -0,0 +1,308 @@
...
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+var (
+	errTest = errors.New("test Error")
+)
+
+func TestBatchCreateCRDs(t *testing.T) {
```
Comment (maxsokolovsky) [inline pkg/crd/crd_test.go:24]: This test and its subtests could also be run in parallel.

File: pkg/yaml/yaml.go (hunk 13 above)
Comment (KevinJoiner) [reply pkg/yaml/yaml.go:35]: The main difference is the new one works with YAML, which contains a list of docs or a single doc. `Unmarshal(...)` only works for bytes that contain a single document. The other function that already exists to help with lists is `ToObject(...)`, but this function forces you to use an `unstructured.Unstructured` objects which as you know cause a lot of headaches when you try to convert them to actual objects with map decode.

`Unmarshal(...)` is still safe if you know your YAML file is not a list of objects. `UnmarshalWithJSONDecoder(...)` provides more functionality, I don't want to depreciate the function since if you are using it today there is no reason to change.

File: pkg/crd/crd.go (hunk 16 above)
Comment (KevinJoiner) [reply pkg/crd/crd.go:19]: Oh nice, I didn't think it would.  I am going to leave it since it is not exported, and it keeps the test from taking a minute to run when testing the public `BatchCreate(..)` function.

Comment (maxsokolovsky) [reply pkg/crd/crd.go:19]: I was thinking you could pass it as a parameter, rather than use the package variable, but it's up to you.

File: pkg/crd/init.go [hunk 19]
```diff
@@ -361,6 +363,35 @@ func (c CRD) ToCustomResourceDefinitionV1Beta1() (*apiextv1beta1.CustomResourceD
 	return v1beta1CRD, nil
 }
 
+// Name resolves the Name for the give CRD.
+func (c *CRD) Name() string {
```
[the remark this answers is not in the document]
Comment (KevinJoiner) [reply pkg/crd/init.go:367]: I just updated the method to do less.

File: pkg/crd/crd.go (hunk 16 above)
Comment (KevinJoiner) [reply pkg/crd/crd.go]: Oh, you mean to the `BatchCreateCRDs()` sure I am fine with that.

Comment (maxsokolovsky) [reply pkg/crd/crd.go]: Or rather `waitCRD` could take it. And `BatchCreateCRDs` would pass it...

File: pkg/crd/mockCRDClient_test.go (hunk 17 above)
Comment (KevinJoiner) [reply pkg/crd/mockCRDClient_test.go:19]: The fake.SimpleClient does not allow me to track what methods were called and with what parameters, the same way mocks do. My first implementation used the fake mocks from K8s but I ended up having to add custom tracking code and the test were pretty hard to follow.

File: pkg/crd/init.go [hunk 20]
```diff
@@ -361,6 +363,31 @@ func (c CRD) ToCustomResourceDefinitionV1Beta1() (*apiextv1beta1.CustomResourceD
 	return v1beta1CRD, nil
 }
 
+// Name resolves the Name for the give CRD.
```
Comment (JonCrowther) [inline pkg/crd/init.go:366]:
```suggestion
// Name resolves the Name for the given CRD.
```

**Review by JonCrowther — APPROVED** [review]

**Review by maxsokolovsky — APPROVED** [review]
